### PR TITLE
Implement SGVector slicing

### PR DESCRIPTION
**Pull request by theartful**
This will be useful for replacing raw pointers. A slice can be reference counted, and we keep track of an offset, which is needed in deallocation.

Why this is useful?
When we want to do an operation with only a part of a vector. For example
```C++
float64_t CCombinedDotFeatures::dense_dot(int32_t vec_idx1, const float64_t* vec2, int32_t vec2_len) const
{
	float64_t result=0;

	uint32_t offs=0;

	for (index_t f_idx=0; f_idx<get_num_feature_obj(); f_idx++)
	{
		CDotFeatures* f = get_feature_obj(f_idx);
		int32_t dim = f->get_dim_feature_space();
		result += f->dense_dot(vec_idx1, vec2+offs, dim)*f->get_combined_feature_weight();
		offs += dim;

		SG_UNREF(f);
	}

	return result;
}
```
Where we do a dot product with `vec2 + offs`. Without slicing, we would have to create a new SGVector that points to `vec2 + offs` then call `dot` which accept SGVector.

### CODE REVIEW

File: src/shogun/lib/SGVector.h [hunk 1]
```diff
@@ -270,6 +269,37 @@ template<class T> class SGVector : public SGReferencedData
...
+		 * @param h slice end index (exclusive)
+		 */
+		SGVector<T> slice(index_t l, index_t h) const;
+
+		/** Operator overloading for vector slicing in read-only mode
+		 *
+		 * @param l slice start index (inclusive)
+		 * @param h slice end index (exclusive)
+		 */
+		inline const SGVector<T> operator()(index_t l, index_t h) const
```
Comment (karlnapf) [inline src/shogun/lib/SGVector.h]: really nice!

**Comment by theartful** [comment]
@karlnapf It's already implemented in `SGMatrix` with the name `submatrix`, but it's not owning, so no reference counting. @vigsterkr also found that reference counting with views isn't necessary, so I think I should change this to that as well.

**Comment by karlnapf** [comment]
Yes subslices of things usually are not stored. However, things should be unified (no different names/behaviour among SGMatrix and SGVector ....)

**Comment by theartful** [comment]
Now they're same name (slice) and same behaviour (not reference counted). I also removed slice operator overloading in SGVector, since in SGMatrix it is used to access a specific element, so it might cause confusion.

File: src/shogun/lib/SGVector.cpp [hunk 2]
```diff
@@ -309,6 +309,15 @@ void SGVector<T>::resize_vector(int32_t n)
 	vlen=n;
 }
 
+template <class T>
+SGVector<T> SGVector<T>::slice(index_t l, index_t h) const
+{
+	assert_on_cpu();
+	ASSERT(l >= 0 && h <= vlen);
+
+	return SGVector<T>(vector, h - l, l);
```
Comment (gf712) [inline src/shogun/lib/SGVector.cpp]: cool! much simpler this way

**Comment by theartful** [comment]
@karlnapf I think this is ready

File: src/shogun/lib/SGVector.cpp (hunk 2 above)
Comment (karlnapf) [inline src/shogun/lib/SGVector.cpp]: so you do reference counting for the returned vector?

Comment (theartful) [reply src/shogun/lib/SGVector.cpp]: no reference counting

**Pull request merged**